Introduce Slurm compute node priority customization

jdeamicis opened the pull request:
### Description of changes
* Introduce StaticNodePriority and DynamicNodePriority configuration parameters at Compute Resource level.

### Tests
* Describe the automated and/or manual tests executed to validate the patch.
* Describe the added/modified tests.

### References
* https://github.com/aws/aws-parallelcluster-cookbook/pull/2127

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [x] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

jdeamicis committed:
Introduce Slurm compute node weight customization

Introduce StaticNodeWeight and DynamicNodeWeight configuration parameters
at Compute Resource level.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Fix unit test with new node weight configuration

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Add Changelog entry for node weights configuration

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Rename Weight to Priority

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Change default value of DynamicNodePriority to 1000

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Add validator to check custom Slurm node priorities

The validator checks that no Slurm dynamic node priority is equal or lower
than any Slurm static node priority in the same queue.
If this happens, the validator throws a warning.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

NSsirena commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -1959,6 +1962,8 @@ def __init__(
         self.health_checks = health_checks or HealthChecks(implied=True)
         self.custom_slurm_settings = Resource.init_param(custom_slurm_settings, default={})
         self.tags = tags
+        self.static_node_priority = Resource.init_param(static_node_priority, default=1)
+        self.dynamic_node_priority = Resource.init_param(dynamic_node_priority, default=1000)
```
[Nitpick] For coherence we should define these constants in `constants.py` (see DEFAULT_MAX_COUNT a few lines above)

jdeamicis replied:
I thought about it, and I didn't do it for the following reasons:
- there are several other places where the default is simply written in the parameter definition, without defining a corresponding constant (I agree we should have more consistency on this in the code);
- if a default is used only once in the config and is independent from other default values, I believe defining a constant is pointless and also slightly annoying, since it forces you to follow a redirection to understand what the default value is.

NSsirena replied:
ok

NSsirena commented on cli/src/pcluster/validators/slurm_settings_validator.py:
```diff
@@ -144,3 +144,36 @@ def _validate(self, custom_settings: List[Dict], include_file_url: str):
+        bad_static_priorities = {
+            key: value
+            for key, value in static_priorities.items()
+            if value in range(min(dynamic_priorities.values()), max(dynamic_priorities.values()) + 1)
```
These functions are evaluated once for each `static priorities` and below once per `dynamic priority`.
Moreover we aren't interested in the full range but only in on boundary: `max_static` and `min_dynamic`
and we want to ensure that "all static are <= min_dynamic" and "all dynamic are >= max_static".

jdeamicis replied:
Oh, you mean calculating first the ranges and using them "statically" in the if condition?

Honest doubt here: wouldn't Python's interpreter try to optimize something here? But I agree we can be stricter.

jdeamicis replied:
Addressed, and we saved some vertical space in the code :)

NSsirena commented on cli/src/pcluster/validators/slurm_settings_validator.py:
```diff
@@ -144,3 +144,36 @@ def _validate(self, custom_settings: List[Dict], include_file_url: str):
+            self._add_failure(
+                f"Some compute resources in queue {queue_name} have static nodes with higher or equal priority than "
+                f"other dynamic nodes in the same queue. "
+                f"Possible problematic static node priorities are {bad_static_priorities}. "
```
Following the previous comment, we can enrich this message reporting also the boundaries extracted above (max_static and min_dynamic). This would make it easier to review the "offending" values.

jdeamicis replied:
The current message is already providing this information, although not explicitly as you are saying.

What is being printed at the moment are all the values of priority (both static and dynamic) that go against the "soft rule" of having all dynamic priorities higher than the static ones within the same queue. So the max static and min dynamic are already included in the printed values (they will be respectively the last value of static priority and the first value of dynamic priority being printed in the message).

NSsirena replied:
Yes. I saw the examples in the tests.
However if you have a long list of Compute Resources is may be confusing and not immediately obvious which value you have to pick from which list..because you have to mentally rebuild the logic behind the check.

While in the other way you can immediately tell the customers which is the "offending threshold" in both cases.
Also you can avoid defining the full range, since what you need is only the min or the max of the range .

Anyway, I'll approve the PR since it is a sort of Nitpick.

jdeamicis committed:
Small refactoring in SlurmNodePrioritiesWarningValidator

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Improve SlurmNodePrioritiesWarningValidator

Improve the logic of the validator to cover cases where values of dynamic
priority are even lower than the range of static priorities and
values of static priority are even higher than the range of the dynamic
ones.

Small improvement to the validator message.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Refine SlurmNodePrioritiesWarningValidator

Do not consider compute resource with no static nodes or no
dynamic nodes when comparing values of static and dynamic
node priorities.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Define min and max values of node priorities in constants.py

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

NSsirena commented on cli/src/pcluster/validators/slurm_settings_validator.py:
```diff
@@ -144,3 +145,34 @@ def _validate(self, custom_settings: List[Dict], include_file_url: str):
+        dynamic_priorities = {
+            cr.name: cr.dynamic_node_priority for cr in compute_resources if cr.max_count > cr.min_count
+        }
+        max_static = max(static_priorities.values()) if len(static_priorities) > 0 else MIN_SLURM_NODE_PRIORITY - 1
+        min_dynamic = min(dynamic_priorities.values()) if len(dynamic_priorities) > 0 else MAX_SLURM_NODE_PRIORITY + 1
```
Please add a comment that explains that we are initializing those values to `impossible to achieve` thresholds because we want to ensure the the warning is not triggered when either static or dynamic nodes are not present in the configuration.

jdeamicis replied:
Addressed.

jdeamicis committed:
Address review comments

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>
